src/main/java/org/corfudb/infrastructure/Rank.java

```diff
@@ -14,32 +14,34 @@
 @Slf4j
 @ToString
 @AllArgsConstructor
-public class Rank implements Comparable<Rank> {
+public class Rank {
     @Getter
     Long rank;
     @Getter
     UUID clientId;
 
     /**
-     * compares this.rank with other.rank
-     * if equal
-     * compares this.clientId with other.clientId
+     * Compares only the ranks. Does not use clientIds in the comparison.
      *
      * @param other
      * @return
      */
-    @Override
-    public int compareTo(Rank other) {
-        /*
-        ** I spent a fair amount of time trying to figure out what prepare
-        ** with a duplicate rank would sometime succeed.  Here is why.
-        ** TODO: uncomment this code once its purpose is understood.
-        ** The LayoutServerTest.checkPhase1AndPhase2MessagesFromMultipleClients test is broken until this hack is reverted.
+    public boolean lessThanEqualTo(Rank other) {
+        if(other == null)
+            return false;
+        if(this == other)
+            return true;
+        return this.rank.compareTo(other.rank) <=0;
+    }
 
-        if (rank.compareTo(other.getRank()) == 0) {
-            return clientId.compareTo(other.clientId);
-        }
-        */
-        return rank.compareTo(other.getRank());
+    @Override
+    public boolean equals(Object obj) {
+        Rank other = (Rank)obj;
```

Comment:
You should check if obj is instanceOf Rank first, otherwise, you could get a ClassCastException if obj is not of the right type, but equals should return false.

I recommend relying on lombok's @EqualsAndHashCode (or just @Data) to generate the correct implementation.